File: src/problems/day01.rs
use crate::problems::Problem;

pub struct Day01 {
    input: Vec<u64>,
    solution1: u64,
    solution2: u64,
}

impl Day01 {
    pub fn new() -> Day01 {
        Day01 {
            input: Vec::new(),
            solution1: 0,
            solution2: 0,
        }
    }
}

impl Problem for Day01 {
    fn setup(&mut self) {
        // let lines = crate::read_lines("input-data/01-test.txt");
        let lines = crate::read_lines("input-data/01-data.txt");
        for s in lines {
            if s.trim().len() == 0 {
                self.input.push(0);
            } else {
                self.input.push(str::parse(s.as_str()).unwrap());
            }
        }
        self.solution1 = 0;
        self.solution2 = 0;
    }

    fn title(&self) -> String {
        String::from("01 - Calorie Counting")
    }

    fn solve_problem1(&mut self) {
        let mut elve_sum: u64 = 0;
        let mut max_sum: u64 = 0;
        for nr in self.input.iter() {
            if *nr > 0 {
                elve_sum += *nr;
            } else {
                if elve_sum > max_sum {
                    max_sum = elve_sum;
                }
                elve_sum = 0;
            }
        }
        // last one, to collect them all:
        if elve_sum > max_sum {
            max_sum = elve_sum;
        }
        self.solution1 = max_sum;
    }
    fn solve_problem2(&mut self) {
        let mut elve_sum: u64 = 0;
        let mut totals: Vec<u64> = Vec::new();

        for nr in self.input.iter() {
            if *nr > 0 {
                elve_sum += *nr;
            } else {
                totals.push(elve_sum);
                elve_sum = 0;
            }
        }
        // collect last one, too:
        totals.push(elve_sum);


        totals.sort();
        let sum = totals.iter().rev().take(3).sum();
        self.solution2 = sum;
    }

    fn solution_problem1(&self) -> String {
        String::from(format!("{}", self.solution1))
    }

    fn solution_problem2(&self) -> String {
        String::from(format!("{}", self.solution2))
    }
}

File: go/problems/day01.go
package problems

import (
	"fmt"
	"sort"
	"strconv"
	"strings"

	"alexi.ch/aoc/2022/lib"
)

type Day01 struct {
	numbers []uint64
	sums    []uint64
	s1      uint64
	s2      uint64
}

func NewDay01() Day01 {
	return Day01{s1: 0, s2: 0, numbers: make([]uint64, 0), sums: make([]uint64, 0)}
}

func (d *Day01) Title() string {
	return "Day 01 - Calorie Counting"
}

func (d *Day01) Setup() {
	// var lines = lib.ReadLines("data/01-test.txt")
	var lines = lib.ReadLines("data/01-data.txt")
	for _, line := range lines {
		nr, err := strconv.ParseUint(strings.TrimSpace(line), 10, 64)
		if err != nil {
			nr = 0
		}
		d.numbers = append(d.numbers, nr)
	}
	// fmt.Printf("%v\n", d.numbers)
}

func (d *Day01) SolveProblem1() {
	var sums = make([]uint64, 0)
	var localSum uint64 = 0

	for _, nr := range d.numbers {
		if nr > 0 {
			localSum += nr
		} else {
			sums = append(sums, localSum)
			localSum = 0
		}
	}
	max, err := lib.FindMax(sums)
	d.sums = sums
	if err == nil {
		d.s1 = *max
	}
}

func (d *Day01) SolveProblem2() {
	// sums already calculated in solve1
	sort.Slice(d.sums, func(i int, j int) bool {
		return d.sums[i] > d.sums[j]
	})
	d.s2 = lib.Sum(d.sums[:3])
}

func (d *Day01) Solution1() string {
	return fmt.Sprintf("%d", d.s1)
}

func (d *Day01) Solution2() string {
	return fmt.Sprintf("%d", d.s2)
}
